benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/DefaultHubDispatcherBenchmark.cs

```diff
@@ -0,0 +1,200 @@
+
+            _connectionContext.ProtocolReaderWriter = new HubProtocolReaderWriter(hubProtocolMock.Object, dataEncoder);
+        }
+
+        public class NoErrorHubConnectionContext : HubConnectionContext
+        {
+            public NoErrorHubConnectionContext(ConnectionContext connectionContext, TimeSpan keepAliveInterval, ILoggerFactory loggerFactory) : base(connectionContext, keepAliveInterval, loggerFactory)
+            {
+            }
+
+            public override Task WriteAsync(HubMessage message)
+            {
+                if (message is CompletionMessage completionMessage)
+                {
+                    if (!string.IsNullOrEmpty(completionMessage.Error))
+                    {
+                        throw new Exception("Error invoking hub method: " + completionMessage.Error);
+                    }
+                }
+
+                return Task.CompletedTask;
+            }
+        }
+
+        public class TestHub : Hub
+        {
+            private static readonly IObservable<int> ObservableInstance = Observable.Empty<int>();
+            private static readonly ChannelReader<int> ChannelReaderInstance = Channel.CreateUnbounded<int>();
+
+            public void Invocation()
+            {
+            }
+
+            public Task InvocationAsync()
+            {
+                return Task.CompletedTask;
+            }
+
+            public int InvocationReturnValue()
+            {
+                return 1;
+            }
+
+            public Task<int> InvocationReturnAsync()
+            {
+                return Task.FromResult(1);
+            }
+
+            public ValueTask<int> InvocationValueTaskAsync()
+            {
+                return new ValueTask<int>(1);
+            }
+
+            public IObservable<int> StreamObservable()
+            {
+                return ObservableInstance;
+            }
+
+            public Task<IObservable<int>> StreamObservableAsync()
+            {
+                return Task.FromResult(ObservableInstance);
+            }
+
+            public ValueTask<IObservable<int>> StreamObservableValueTaskAsync()
+            {
+                return new ValueTask<IObservable<int>>(ObservableInstance);
+            }
+
+            public ChannelReader<int> StreamChannelReader()
+            {
+                return ChannelReaderInstance;
+            }
+
+            public Task<ChannelReader<int>> StreamChannelReaderAsync()
+            {
+                return Task.FromResult(ChannelReaderInstance);
+            }
+
+            public ValueTask<ChannelReader<int>> StreamChannelReaderValueTaskAsync()
+            {
+                return new ValueTask<ChannelReader<int>>(ChannelReaderInstance);
+            }
+        }
+
+        [Benchmark]
+        public Task Invocation()
+        {
+            return _dispatcher.DispatchMessageAsync(_connectionContext, new InvocationMessage("123", "Invocation", null));
```

Comment:
Benchmarks differ only in one string. Params(https://github.com/dotnet/BenchmarkDotNet/blob/master/docs/guide/Advanced/Params.md) can be used for this purpose.

Reply:
I thought about that but the don't differ by just one string - there are invocation messages and stream messages which have a different message type.